src/threads_arg/normalization.py

```diff
@@ -0,0 +1,122 @@
+# This file is part of the Threads software suite.
+# Copyright (C) 2024-2025 Threads Developers.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import logging
+import pandas as pd
+import math
+import numpy as np
+import pandas
+from threads_arg import ThreadingInstructions
+import msprime
+
+logging.basicConfig(
+    format='%(asctime)s %(levelname)-8s %(message)s',
+    level=logging.INFO,
+    datefmt='%Y-%m-%d %H:%M:%S')
+logging.getLogger("msprime").setLevel(logging.WARNING)
+
+class Normalizer:
+    """
+    Minimal simulator based on ARG-Needle Simulator class
+    and including the ARG-Needle normalisation routine.
+    """
+    def __init__(self, demography_file, num_samples):
+        self.num_samples = num_samples
+        self.demography_file = demography_file
+        self.demography = self.read_demography(demography_file)
+
+    def read_demography(self, demography_file):
+        df = pd.read_table(demography_file, header=None)
+        df.columns  = ['GEN', 'NE']
+        demography = msprime.Demography()
+        # NOTE: these initial sizes get overwritten anyways
+        demography.add_population(name="A", initial_size=1e4)
+        for t, ne in zip(df["GEN"], df["NE"]):
+            # Divide by 2 because msprime wants diploids but demography is haploid
+            demography.add_population_parameters_change(t, initial_size=ne / 2, population="A")
+        return demography
+
+    def simulation(self, length, random_seed=10):
+        ts = msprime.sim_ancestry(
+            samples={"A": self.num_samples // 2},  # again, diploid
+            demography=self.demography,
+            sequence_length=length,
+            recombination_rate=0,
+            random_seed=random_seed)
+        return ts
+
+    def normalize(self, threading_instructions, num_seeds=1000, start_seed=1):
+        num_samples = threading_instructions.num_samples
+        assert num_samples == self.num_samples
+        
+        heights_to_span = {}
+        positions = threading_instructions.positions
+
+        for start_vec, tmrca_vec in zip(threading_instructions.all_starts(), threading_instructions.all_tmrcas()):
+            assert len(start_vec) == len(tmrca_vec)
+            for i, (start, tmrca) in enumerate(zip(start_vec, tmrca_vec)):
+                try:
```

Comment:
Optional: Would this be clearer with an `(i + 1) in start_vec` condition rather than exception?

* I've just read that exceptions are not considered expensive in python (compared to C++, say) so perhaps this is idiomatic code, leaving this as an optional.